### ISSUE TITLE
Debug assertion failure: `is_soft_unit_vector(init.dir)`

### ISSUE THREAD

**Issue by sethrj** [celeritas-project/celeritas#639]
@amandalund discovered that ORANGE geometries are now failing after #625 fixed positron annihilation: secondaries being created *only on GPU* trigger an assertion:
```
/home/s3j/.local/src/celeritas/src/orange/OrangeTrackView.hh:236:
celeritas: internal assertion failed: is_soft_unit_vector(init.dir)
```

This shows up in basically all the regression problems that use ORANGE. It's not clear why we don't see this in our `app/demo-loop` test.

Because v0.2.0 doesn't include positron annihilation (whoops!) we don't see the problem. The `backports/v0.1` branch, which was before the recent multi-universe work on ORANGE but includes the annihilation fixes, does *not* show the problem.

**Comment by amandalund** [comment]
Actually I do also see it in the `app/demo-loop` GPU test (with debug on and reldeb build).

**Comment by sethrj** [comment]
I wonder if it's something to do with the "let's help our debugging" implementation when resizing states:
```c++
template<class T, MemSpace M, class I>
void CollectionBuilder<T, M, I>::resize(size_type size)
{
    CELER_EXPECT(this->storage().empty());
    this->storage() = StorageT(size);
    if constexpr (CELERITAS_DEBUG && M == MemSpace::host)
    {
        // Fill with invalid values to help with debugging on host
        detail::fill_invalid(&col_);
    }
}
```

Either way, we clearly need to add a reldeb configuration to our CI after this...

**Comment by sethrj** [comment]
Necessary conditions via @amandalund :
-`CELERITAS_DEBUG=true`
- release or reldeb build (some optimizations)
- GPU only, not CPU
- positron annihilation enabled

directions coming out of interaction are good; but when trying to create secondaries the directions have been zeroed out

**Comment by amandalund** [comment]
One more observation: running with a single thread on the GPU it fails with a different assertion:
```console
/home/alund/celeritas_project/celeritas/src/corecel/data/Collection.hh:367:
celeritas: internal assertion failed: i < this->size()
```
I traced the failure back to [this line](https://github.com/celeritas-project/celeritas/blob/0e3b0d511dfb87e12e434a463e26e4a91edb3482/src/corecel/data/StackAllocator.hh#L231) in the`StackAllocator`. Immediately before [this line](https://github.com/celeritas-project/celeritas/blob/0e3b0d511dfb87e12e434a463e26e4a91edb3482/src/corecel/data/StackAllocator.hh#L227) the local variable `start` has the correct value, and immediately after it has a bogus value.

**Comment by amandalund** [comment]
> I wonder if it's something to do with the "let's help our debugging" implementation when resizing states:
> 
> ```c++
> template<class T, MemSpace M, class I>
> void CollectionBuilder<T, M, I>::resize(size_type size)
> {
>     CELER_EXPECT(this->storage().empty());
>     this->storage() = StorageT(size);
>     if constexpr (CELERITAS_DEBUG && M == MemSpace::host)
>     {
>         // Fill with invalid values to help with debugging on host
>         detail::fill_invalid(&col_);
>     }
> }
> ```
Removing this block had no effect.

**Comment by sethrj** [comment]
If this *is* a weird stack overflow like you think @amandalund , we could try updating the stack size like #638 as well?

**Comment by amandalund** [comment]
Not sure if it's an overflow, but it does seem like something is corrupting the stack. I did try increasing the cuda stack size at some point and it didn't seem to help..

**Comment by sethrj** [comment]
We think this is a weird compiler bug rather than anything wrong on our side. Depending on minor irrelevant changes to ORANGE, this sometimes works and sometimes doesn't.

**Comment by amandalund** [comment]
I'm not totally convinced this one was a compiler bug, but I suppose we can deal with it if/when it reappears.

**Comment by amandalund** [comment]
> One more observation: running with a single thread on the GPU it fails with a different assertion:
> 
> ```
> /home/alund/celeritas_project/celeritas/src/corecel/data/Collection.hh:367:
> celeritas: internal assertion failed: i < this->size()
> ```
> 
> I traced the failure back to [this line](https://github.com/celeritas-project/celeritas/blob/0e3b0d511dfb87e12e434a463e26e4a91edb3482/src/corecel/data/StackAllocator.hh#L231) in the`StackAllocator`. Immediately before [this line](https://github.com/celeritas-project/celeritas/blob/0e3b0d511dfb87e12e434a463e26e4a91edb3482/src/corecel/data/StackAllocator.hh#L227) the local variable `start` has the correct value, and immediately after it has a bogus value.

Still hitting this sometimes on the current develop with:

- `CELERITAS_DEBUG=true`
- reldeb (but not debug) build
- GPU runs only
- VecGeom this time

Running regression problems with `max_num_tracks = 1`, e.g.
<details><summary>simple-cms JSON input</summary>

```json
{
"_geometry": "vecgeom",
"_instance": 1,
"_name": [
"simple-cms+field+msc",
"vecgeom",
"gpu"
],
"_num_events": 7,
"_num_primaries": 9100,
"_outdir": "simple-cms+field+msc-vecgeom-gpu",
"_timeout": 600.0,
"brem_combined": false,
"enable_diagnostics": false,
"geant_options": {
"apply_cuts": true,
"coulomb_scattering": false,
"eloss_fluctuation": false,
"em_bins_per_decade": 56,
"lpm": true,
"msc": "urban",
"physics": "em_basic",
"rayleigh_scattering": true
},
"geometry_filename": "/home/alund/celeritas_project/regression/input/simple-cms.gdml",
"hepmc3_filename": "/home/alund/celeritas_project/regression/input/simple-cms-13TeV.hepmc3",
"initializer_capacity": 67108864,
"mag_field": [
0.0,
0.0,
1.0
],
"max_events": 7,
"max_num_tracks": 1,
"physics_filename": "/home/alund/celeritas_project/regression/input/simple-cms.gdml",
"secondary_stack_factor": 3.0,
"seed": 20220906,
"sync": true,
"use_device": true
}
```

</details>

with a couple `printf`s added to the `StackAllocator`

<details open><summary>diff</summary>

```diff
diff --git a/src/corecel/data/StackAllocator.hh b/src/corecel/data/StackAllocator.hh
index bcca60b4..9e5f7014 100644
--- a/src/corecel/data/StackAllocator.hh
+++ b/src/corecel/data/StackAllocator.hh
@@ -224,10 +224,12 @@ CELER_FUNCTION auto StackAllocator<T>::operator()(size_type count)
     }
 
     // Initialize the data at the newly "allocated" address
+    printf("start: %u\n", start);
     value_type* result = new (&data_.storage[StorageId{start}]) value_type;
     for (size_type i = 1; i < count; ++i)
     {
         // Initialize remaining values
+        printf("i: %u, start: %u\n", i, start);
         new (&data_.storage[StorageId{start + i}]) value_type;
     }
     return result;
```

</details>

produces the output:

<details open><summary>output</summary>

```console
$ ./demo-loop simple-cms+field+msc-vecgeom-gpu.inp.json 
status: Loading input and initializing problem data
status: Initializing Geant4 run manager
status: Initializing Geant4 geometry and physics
info: Loading Geant4 geometry from GDML at /home/alund/celeritas_project/regression/input/simple-cms.gdml
status: Initializing Geant4 physics tables
status: Transferring data from Geant4
status: Loading external elemental data
info: Loading VecGeom geometry from GDML at /home/alund/celeritas_project/regression/input/simple-cms.gdml
status: Initializing tracking information
status: Reading and building Seltzer Berger model data
status: Reading and building Livermore PE model data
status: Transporting
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
start: 0
i: 1, start: 0
start: 0
start: 0
start: 0
i: 1, start: 4
/home/alund/celeritas_project/celeritas/src/corecel/data/Collection.hh:414:
celeritas: internal assertion failed: i < this->size()
terminate called after throwing an instance of 'celeritas::RuntimeError'
  what():  /home/alund/celeritas_project/celeritas/src/corecel/data/DeviceAllocation.cc:61:
celeritas: CUDA error: cudaFree(ptr) failed:
    unspecified launch failure
Aborted
```

</details>

**Comment by sethrj** [comment]
@amandalund Is this reasonably reproducible? Maybe it's time to bring an NVIDIA engineer in on this...

**Comment by sethrj** [comment]
From @amandalund on slack:
> I also confirmed the LivermorePE inc_energy_ > 0 failure is the same behavior we were seeing in #639: secondaries are created in an interaction with valid direction/energy, but immediately after the Interactor launch some are zeroed out (by other threads initializing secondaries in memory that doesn’t belong to them, because the start variable is somehow changing in the StackAllocator)

See #715

**Comment by sethrj** [comment]
From @amandalund on slack:
> I was able to test out copying the interactor kernel arguments to the device and passing a pointer to them instead, and it *did* get rid of the weird stack corruption. I’m not quite sure what to make of that because the size of the arguments is ~1.2 KB, which should be well below the limit (from the CUDA C++ Programming Guide):
>> __global__ function parameters are passed to the device via constant memory and are limited to 32,764 bytes starting with Volta, and 4 KB on older architectures.

> and because it’s not the only way to fix (hide? work around?) the bug; just as an example, this also gets rid of it:
```diff
diff --git a/src/corecel/Types.hh b/src/corecel/Types.hh
index 84224e2b..126ac19a 100644
--- a/src/corecel/Types.hh
+++ b/src/corecel/Types.hh
@@ -17,7 +17,7 @@ namespace celeritas
 //---------------------------------------------------------------------------//
 #if CELER_USE_DEVICE
 //! Standard type for container sizes, optimized for GPU use.
-using size_type = unsigned int;
+using size_type = unsigned long long int;
 #else
 using size_type = std::size_t;
 #endif
```

**Comment by sethrj** [comment]
Our best guess is that this is a compiler bug caused y having too much "constant" memory. It seems to have been fixed by #741 since #715 passes on the CI now...
